[Feature/311] 모임 기록 중 활동을 생성하는 API 구현

hosunglee222 opened the pull request:
## Type of change
<!-- 작업의 종류를 선택해주세요. -->
- [X] Feature : 새로운 기능 추가
- [ ] Bug fix : 버그 수정
- [ ] Refactor : 코드 리팩토링 작업
- [ ] Document : 문서작업
- [ ] Test : 테스트 코드 작성 및 테스트 작업
- [ ] Style : 코드 스타일 및 포맷팅 작업
- [ ] CI/CD : CI/CD 작업 수정
- [ ] Chore : 패키지 매니저, 라이브러리 업데이트 등의 작업

## PR Desciption

> 변경 사항 설명

- 활동을 생성
  - 활동 참여자 정보 생성
  - 활동 정산 정보 생성

## Requirements for Reviewer
> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
> ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요?

- 활동 생성에 필요한 값 확인 부탁드립니다.

## PR Log

> PR 작업하면서 고민했던 내용, 해결한 내용, 고민 중인 내용 등

### 새롭게 배운 것

- 

### 고민 중인 사항

- 현재 활동 생성 API 자체가 매우 방대하여 UI 변경시 API를 분리 할 필요성이 느껴집니다.

## 첨부 자료
<img width="721" alt="image" src="https://github.com/user-attachments/assets/c5694f28-86e3-4e7e-b371-a197af778a30">


## 관련 이슈

- close #311

hosung-222 committed:
:sparkles: Feature: 활동 생성 API 스펙 명세

<body>
그룹 스케줄의 기록 중 활동을 생성하는 API 스펙 명세

- 관련 : #311

hosung-222 committed:
:sparkles: Feature: 활동 엔티티에 위치 정보 추가

<body>
활동의 위치를 기록하기 위한 위치 정보 추가

- 관련 : #311

hosung-222 committed:
:sparkles: Feature: ActivityConverter 구현

<body>
Request dto -> Entity

- 관련 : #311

hosung-222 committed:
:sparkles: Feature: 활동 정산 내역 추가하는 메서드 구현

<body>
정산 내역 Set

- 관련 : #311

hosung-222 committed:
:sparkles: Feature: Participantid로 찾는 메서드 구현

- 관련 : #311

hosung-222 committed:
:sparkles: Feature: 활동 참여 내역 생성하고 정산 추가하는 메서드 구현

<body>
활동 참여 내역을 추가하고 정산 내역을 추가하는 메서드 구현

- 관련 : #311

hosung-222 committed:
:sparkles: Feature: 활동 기록에 이미지 값 추가

<body>
이미지 값 추가하도록 구현

- 관련 : #311

hosung-222 committed:
:sparkles: Feature: 활동 생성 비즈니스 로직 구현

<body>
활동, 활동 이미지, 참여정보 등 생성

- 관련 : #311

hosung-222 committed:
:sparkles: Feature: 활동 생성 API 구현 완료

<body>
활동생성 API 구현 완료

- 관련 : #311

hosung-222 committed:
:bug: Fix: 활동이 생성되기 이전에 참조되는 버그 해결

<body>
활동생성 순서 변경

- 관련 : #311

hosung-222 committed:
:bug: Fix: 활동 시작 끝나는일을 가져오도록 수정

<body>
기존 스케줄 시간 -> 활동 별 시간으로 수정

- 관련 : #311

hosung-222 committed:
:bug: Fix: 활동 시작 끝나는일 저장하도록 수정

<body>
활동 일정 기록되도록 수정

- 관련 : #311

hosung-222 committed:
:memo: Docs: 스웨거 명세 추가

<body>
dto 스웨거 명세 추가

- 관련 : #311

hosung-222 committed:
:memo: Docs: 주석 추가

<body>
usecase 주석 추가

- 관련 : #311

joowojr commented on application/external-api-v2/src/main/java/com/namo/spring/application/external/api/record/usecase/ActivityUseCase.java:
```diff
@@ -36,4 +40,18 @@ public ActivityResponse.ActivitySettlementInfoDto getSettlement(Long memberId, L
+        // 활동 생성
+        Activity activity = activityManageService.createActivity(memberId, scheduleId, request);
+        // 활동 참여 정보 생성
+        List<ActivityParticipant> activityParticipant = activityParticipantManageService.createActivityParticipant(
```
participant ID 배열을 검증하는 로직을 추가하면 더 좋을것 같습니다. 
ex) MeetingParticipantValidationUtils class의 validateUniqueParticipantId 참고

hosunglee222 replied:
오호 감사합니다!

joowojr commented on application/external-api-v2/src/main/java/com/namo/spring/application/external/api/record/serivce/ActivityParticipantManageService.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    public List<ActivityParticipant> createActivityParticipant(Activity activity, List<Long> participantIdList) {
+        List<Participant> participants = participantIdList.stream()
+                .map(participantManageService::getParticipant)
```
participant id list 로 한 번에 조회한 뒤 list의 size를 비교하여 유효성 검증을 하는 방법은 어떨까요?

hosung-222 replied:
요청 받은 request Id list 값과 발견한 참여 기록의 개수 비교 검증을 말씀하시는게 맞을까요? 
좋을 것 같습니다!

hosunglee222 replied:
요청 받은 request Id list 값과 발견한 참여 기록의 개수 비교 검증을 말씀하시는게 맞을까요?
좋을 것 같습니다!

hosunglee222 replied:
로직을 다시 살펴보니 조회할때 Participant가 존재하지 않으면 예외처리되고 트랜잭션이롤백되고 API가 실패하는데 해당 validation이 필요할까요?

joowojr commented on application/external-api-v2/src/main/java/com/namo/spring/application/external/api/record/serivce/ActivityManageService.java:
```diff
@@ -27,4 +34,23 @@ public Activity getMyActivity(Long memberId, Long activityId) {
+        Participant myParticipant = participantManageService.getMyParticipant(memberId, scheduleId);
+        Activity activity = activityService.createActivity(ActivityConverter.toActivity(myParticipant.getSchedule(), request));
+        // 활동 이미지 생성
+        if (request.getImageList() != null && !request.getImageList().isEmpty()){
```
saveAll로 한 번에 이미지를 생성하는 방식은 어떨까요?

hosunglee222 replied:
추후 이미지 수정을 위해 하나만 생성후 저장하는 로직 구현시 유연하게 사용하기 위해 service단에서 하나씩 생성 저장하고 밖에서 forEach를 사용했습니다.
flush가 되지 않아 DB로 save 쿼리가 한번에 될 것 같아서 문제가 없을 것 같아 해당 방식으로 처리했는데 성능 이슈를 확인해보고 문제가 있으면 수정하겠습니다.

joowojr commented on application/external-api-v2/src/main/java/com/namo/spring/application/external/api/record/serivce/ActivityImageManageService.java:
```diff
@@ -0,0 +1,25 @@
+
+    private final ActivityImageService activityImageService;
+
+    public void createActiveImage(Activity activity, String image) {
```
imageOrder가 0인 이유가 있을까요?

hosunglee222 replied:
현재 기획과 클라이언트에서 이미지 순서를 사용하지 않는다고 하여 0으로 우선 사용하려 합니다.

joowojr reviewed (approved):
수고하셨습니다. 리뷰 확인 부탁드립니다 :)

joowojr commented on application/external-api-v2/src/main/java/com/namo/spring/application/external/api/record/dto/ActivityRequest.java:
```diff
@@ -0,0 +1,57 @@
+    public static class CreateActivityDto{
+        @Schema(description = "활동 제목", example = "보드게임 밥디")
+        private String title;
+        @Schema(description = "활동에 참여할 participantId 를 넣어주세요", example = "[1, 2, 3]")
```
request에서 list로 받는 값들은 null exception 방지를 위해 @NotNull로 지정하는 것은 어떨까요?

hosunglee222 replied:
좋습니다! 수정하겠습니다.

hosung-222 committed:
:sparkles: Feat: 활동 참여자 id 중복 요청 검증

<body>
validateDuplicateParticipantIds() 구현

- 관련 : #311

hosung-222 committed:
:sparkles: Feat: 새로운 예외처리 용 에러 코드 구현

<body>
NOT_SCHEDULE_PARTICIPANT_OR_NOT_ACTIVE : 해당 일정의 참여자가 아닌 사람이 포함되어있거나 스케줄 초대를 수락하지 않은 사람이 포함되어있는 경우

- 관련 : #311

hosung-222 committed:
:sparkles: Feat: 참여자 ID List와 스케줄 ID를 이용해 참여자 List 반환하는 도메인 로직구현

<body>
참여자 ID List와 스케줄 ID를 이용해 스케줄에 참여에 수락한 참여자 List 반환하는 도메인 로직구현

- 관련 : #311

hosung-222 committed:
:sparkles: Feat: 참여자 ID List와 스케줄 ID를 이용해 참여자 List 반환하는 도메인 로직구현

<body>
참여자 ID List와 스케줄 ID를 이용해 스케줄에 참여에 수락한 참여자 List 반환하는 도메인 로직구현

- 관련 : #311

hosung-222 committed:
:recycle: Refactor: 활동 참여자 생성시 검증 로직 추가

<body>
활동 참여자 중복 여부, 활동 참여자가 같은 스케줄의 포함되어있는 지 검증 (만약 초대받았어도 거절했거나, 수락하지 않았다면 활동에 포함되지 못함)

- 관련 : #311

hosung-222 committed:
:recycle: Refactor: 활동 생성 API 수정

<body>
- 검증 로직 추가
- 스케줄 참여 수락한 맴버만 활동 참여 가능하도록 변경

- 관련 : #311

hosung-222 committed:
:recycle: Refactor: 활동 이미지 생성 로직 리팩토링

<body>
- 활동 이미지를 생성후 한번에 saveAll하도록 변경

- 관련 : #311

hosung-222 committed:
:recycle: Refactor: DTO nullalble 조건추가

<body>
- 활동 제목과 참여자에대해 Nullable 추가

- 관련 : #311

hosung-222 committed:
:recycle: Refactor: DTO nullalble 조건추가

<body>
- 활동 제목과 참여자에대해 Nullable 추가

- 관련 : #311

joowojr committed:
:sparkles: Feat : activity 테이블에 location, start_date, end_date 컬럼 추가

<footer>
- 관련: #311

joowojr committed:
:art: Style : dto validation 추가

<footer>
- 관련: #311
